src/access_control.rs

```diff
@@ -0,0 +1,43 @@
+use std::marker::PhantomData;
+
+/// Access control policy used to check if a domain is allowed in
+/// either the `Host` or `Origin` header.
+pub trait Policy {
+    fn is_allowed(&self, domain: &[u8]) -> bool;
+}
+
+/// Allow any domain, implements [`Policy`].
+#[derive(Debug)]
+pub struct AllowAny;
+
+impl Policy for AllowAny {
+    fn is_allowed(&self, _domain: &[u8]) -> bool {
+        true
+    }
+}
+
+/// Allow only domains from the list, implements [`Policy`].
+#[derive(Debug)]
+pub struct AllowList<List, Domain> {
```

Comment:
So here `List` can be anything that is iterable, neat.

Reply:
everything that is iterable with `Item = AsRef<str>` right?

Reply:
Anything that can be compared I think (`Eq`)? The `Domain` type param can be anything you want (defaults to `str` though).